community/cypher/cypher-compiler-3.3/src/main/scala/org/neo4j/cypher/internal/compiler/v3_3/planner/logical/CardinalityCostModel.scala

```diff
@@ -117,7 +117,9 @@ case class CardinalityCostModel(config: CypherCompilerConfiguration) extends Cos
   def apply(plan: LogicalPlan, input: QueryGraphSolverInput): Cost = {
     val cost = plan match {
       case CartesianProduct(lhs, rhs) =>
-        apply(lhs, input) + lhs.solved.estimatedCardinality * apply(rhs, input)
+        val lhsCardinality = Cardinality.max(Cardinality.SINGLE, lhs.solved.estimatedCardinality)
```

Comment:
Maybe its not the scope of this PR, but should we also do this for HashJoin below?

Reply:
I don't think we should. The join calculation does not hide the cost of either child no matter what the cardinalities are - that is the only problem this PR is trying to fix.